start_admin_sshd.sh

```diff
@@ -6,38 +6,94 @@ set -e
 
 mkdir -p /home/ec2-user/.ssh/
 chmod 700 /home/ec2-user/.ssh/
-ssh_host_key_dir="/.bottlerocket/host-containers/admin/etc/ssh"
-ssh_config_dir="/home/ec2-user/.ssh"
-
-# Populate authorized_keys with all the public keys found in instance meta-data
-# The URLs for keys include an index and the keypair name, e.g.
-# http://169.254.169.254/latest/meta-data/public-keys/0=mykeypair/openssh-key
-ssh_authorized_keys="${ssh_config_dir}/authorized_keys"
-touch ${ssh_authorized_keys}
-chmod 600 ${ssh_authorized_keys}
-public_key_base_url="http://169.254.169.254/latest/meta-data/public-keys/"
-imds_session_token=$(curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 60")
-imds_request_add_header="X-aws-ec2-metadata-token: ${imds_session_token}"
-public_key_indexes=($(curl -H "${imds_request_add_header}" -sf "${public_key_base_url}" \
-    | cut -d= -f1 \
-    | xargs))
-
-for public_key_index in "${public_key_indexes[@]}"; do
-  public_key_data="$(curl -H "${imds_request_add_header}" -sf "${public_key_base_url}/${public_key_index}/openssh-key")"
-  if [[ ! "${public_key_data}" =~ ^"ssh" ]]; then
-    echo "Key ${public_key_data} with index ${public_key_index} looks invalid" >&2
-    continue
-  fi
-  echo "${public_key_data}" >> "${ssh_authorized_keys}"
-  if ! grep -q "${public_key_data}" "${ssh_authorized_keys}"; then
-    echo "Failed to write key with index ${public_key_index} to authorized_keys" >&2
-    continue
+declare -r ssh_host_key_dir="/.bottlerocket/host-containers/admin/etc/ssh"
+declare -r ssh_config_dir="/home/ec2-user/.ssh"
+declare -r user_data="/.bottlerocket/host-containers/admin/user-data"
+available_auth_methods=0
+
+parse_authorized_keys() {
+    # Extract the keys from user-data json
+    local raw_keys
+    if ! raw_keys=$(jq -e -r '.["ssh"]["authorized_keys"][]' "${user_data}" 2>/dev/null); then
+      echo "Failed to parse authorized_keys from ${user_data}" >&2
```

Comment:
The `echo msg >&2` is repeated throughout, instead of relying on the convention of redirection a function would make the intent clear and safer overall. It would reduce the chances that extra output is sent to subshell substitutes that are expecting data in stdout.

```bash
log() {
  echo "$*" >&2
}
```

```suggestion
      log "Failed to parse authorized_keys from ${user_data}"
```


<details>
<summary>Or, alternatively (using my go to <samp>log</samp> definition)</summary>

```bash
# log with date prefix. 
# 
# Suggested use: `log $LEVEL: $msg`
log() {
  echo "[$(date --iso-8601=s --utc)] $*" >&2
}

# with use as:

log "INFO: installed ssh authorized keys"
log "WARN: unable to parse trusted user CA certificates"
log "ERROR: unable to setup ssh with provided user data"
```

</details>